larcore/Geometry/GeometryConfigurationWriter_service.cc

```diff
@@ -0,0 +1,254 @@
+ * `sumdata::GeometryConfigurationInfo` is present into the run, accessible with
+ * an input tag `GeometryConfigurationWriter`:
+ * 
+ * * if such information is already available in the run, no further information
+ *   is added
+ * * legacy: if there is no such information, but there is a `sumdata::RunData`
+ *   data product, a reduced version of the configuration information is created
+ *   from the information in that data product (the first one, if multiple are
+ *   present)
+ * * finally, if no information is present neither in the full 
+ *   `sumdata::GeometryConfigurationInfo` form nor in the legacy
+ *   `sumdata::RunData` form, information is put together based on the current
+ *   configuration of the `Geometry` service.
+ * 
+ * 
+ * Service dependencies
+ * ---------------------
+ * 
+ * * `Geometry` service
+ *   (for obtaining the current configuration to put into the event)
+ * 
+ */
+class geo::GeometryConfigurationWriter: public art::ProducingService {
+  
+    public:
+  
+  /// Service configuration.
+  struct Config {
+    // no configuration parameters at this time
+  };
+  
+  using Parameters = art::ServiceTable<Config>;
+  
+  /// Constructor: gets its configuration and does nothing with it.
+  GeometryConfigurationWriter(Parameters const&);
+  
+  
+    private:
+  
+  /// Alias for the pointer to the data product object to be put into the run.
+  using InfoPtr_t = std::unique_ptr<sumdata::GeometryConfigurationInfo>;
+  
+  
+  // -- BEGIN --- Interface ----------------------------------------------------
+      
+  /// Writes the information from the service configuration into the `run`.
+  virtual void postReadRun(art::Run& run) override;
+  
+  // -- END --- Interface ------------------------------------------------------
+  
+  
+  /// Loads the geometry information from the `run` (either directly or legacy).
+  InfoPtr_t loadInfo(art::Run& run) const;
+  
+  /// Creates configuration information based on the current `Geometry` service.
+  static InfoPtr_t extractInfoFromGeometry();
+  
+  /// Reads geometry information from the run (returns null pointer if none).
+  InfoPtr_t readGeometryInformation(art::Run& run) const;
+  
+  /// Upgrades legacy `sumdata::RunData` in `run` to geometry information
+  /// (returns null pointer if no legacy information is present).
+  InfoPtr_t makeInfoFromRunData(art::Run& run) const;
+  
+  /// Returns a pointer to the `sumdata::RunData` in `run` (nullptr if none).
+  sumdata::RunData const* readRunData(art::Run& run) const;
+  
+  /// Converts the legacy `data` into geometry configuration information.
+  static InfoPtr_t convertRunDataToGeometryInformation
+    (sumdata::RunData const& data);
+  
+  /// Alias to `std::make_unique<sumdata::GeometryConfigurationInfo>`
+  template <typename... Args>
+  static InfoPtr_t makeInfoPtr(Args&&... args)
+    {
+      return std::make_unique<sumdata::GeometryConfigurationInfo>
+        (std::forward<Args>(args)...);
+    }
```

Comment:
Suggest making this a normal (perhaps `static`) member function--all uses of it below pass one argument and it's always of the same type.  Avoiding unnecessary templates prevents unnecessary compilation time.

Reply:
It is not exactly true that the type in the three calls is the same: in one case is a l-value `const` reference, and in the other two it is a r-value one.
We are hitting the border of my understanding of C++ here. Is it correct that the same effect can be obtained by passing the argument by value?
```
static InfoPtr_t makeInfoPtr(sumdata::GeometryConfigurationInfo info)
    {
      return std::make_unique<sumdata::GeometryConfigurationInfo>
        (std::move(info));
    }
```
In my understanding, passing a r-value reference would move its content into `info`, and passing a l-value reference would copy it.

Reply:
You're not getting a lot of savings by using `std::move` on a `GeometryConfigurationInfo` object--it contains a number and two strings.  So in this case, I wouldn't worry about eking out those 10s of bytes saved by using the move in one case and the copy in the other.  The code will be easier to read, and almost as performance to just do:
```.cpp
static InfoPtr_t makInfoPtr(sumdata::GeometryConfiguationInfo const& info)
{
  return std::make_unique<sumdata::GeometryConfigurationInfo>(info);
}
```
In fact, it may be _just as_ performant as you might be copying the stack's info object to the heap anyway (depending on the size of the internal strings).